lib/plugins/threaded-signing/plugin.c

```diff
@@ -155,17 +161,19 @@ signing_worker_thread(void *user_data)
   while (self->is_running) {
     if (self->input_buffer_idx > 0) {
       // Get the oldest hash from the input buffer
-      assert(input_buffer_end <= MAX_BUFFER_LENGTH);
+      assert(self->input_buffer_idx <= MAX_BUFFER_LENGTH);
       // Copy the hash to |signature_info| and start signing. In principle, it is now possible to
       // prepare for a new hash.
       assert(self->hash_size == self->signature_info->hash_size);
       assert(self->signature_info->hash);
       memcpy(self->signature_info->hash, self->input_buffer[0], self->hash_size);
 
-      uint8_t * tmp = self->input_buffer[0];
+      uint8_t *tmp = self->input_buffer[0];
       int j = 0;
-      while (self->input_buffer[j + 1] != NULL) {
-        self->input_buffer[j]= self->input_buffer[j + 1];
+      while (self->input_buffer[j] != NULL) {
+        self->input_buffer[j] = self->input_buffer[j + 1];
+        if (self->input_buffer[j + 1] != NULL) {
+        }
         j++;
       }
       self->input_buffer[j + 1] = tmp;
```

Comment:
This will create a gap in the buffer.
Consider the buffer
```
[0xabc, 0x0, 0x0, 0x0]
```
then ```tmp = 0xabc``` and the after the while loop you have
```
[0x0, 0x0, 0x0, 0x0]
j = 1
```
so after ```self->input_buffer[j+1] = self->input_buffer[2] = tmp = 0xabc```
hence
```
[0x0, 0x0, 0xabc, 0x0]
```